### PR TITLE
Feature/source site polymorphism

### PR DESCRIPTION
**Pull request by anaximander23**
Addresses #58 
Lays some groundwork for #47 
Should streamline the process of adding new supported sources for minis.

### CODE REVIEW

File: MiniIndex/Pages/Creators/Details.cshtml [hunk 1]
```diff
@@ -7,127 +7,74 @@
...
                         </div>
-                    }
-                </div>
+                    </div>
+                @: </div>
+            }
             </div>
         </div>
-    </div>
-    <div class="tab-pane" id="thingiverse" role="tabpanel" aria-labelledby="thingiverse-tab">
```
Comment (aluhrs13) [inline MiniIndex/Pages/Creators/Details.cshtml]: This merge is wrong, but easy to clean-up afterwards.

**Comment by aluhrs13** [comment]
- [ ] Every Mini added adds a record to SourceSite, regardless of existing SourceSite (test with onmioji's Thingiverse and mz4250's Shapways)
- [ ] Trying to add a Mini from a creator's page (/Minis/Create?URL=<url>) doesn't work.
- [ ] /Minis/Details still uses Mini.Link instead of MiniSourceSite
- [ ] /Minis/Index, /Creators/Details still use Mini.Link instead of MiniSourceSite
- [ ] #-style links from Gumroad not parsing (https://gumroad.com/nafarrate#jJbmb)
- [ ] Double-check migration worked right for Gumroad, creators that previously had a gumroad as their "website" are showing those as website still and could cause issues. Probably manually fix?
- [ ] New Thingiverse Minis are added with an API link.

Works:
- Adding a Mini with an existing creator - Thingiverse, Shapeways, Gumroad  (issues #1 and #5 above)
- Adding an existing Mini - Thingiverse, Shapeways, Gumroad
- Adding a Mini with a new creator - Thingiverse, Gumroad
- Viewing a Mini and opening external link (not using new code though, #3 above)
- Approving a Mini
- Viewing a Creator (minor issue, #2 above)
- Tagging

Polish follow-ups (probably do after merging to not make this PR any bigger):
- Add a dropdown when more than one MiniSourceSite
- Add tests for all parsers
- Making parsing logic more rigid to handle if there's another thing in the query string like a social media link (this was weak before too)

**Comment by aluhrs13** [comment]
Opened #119 for figuring out the future of Mini.Link, #120 to go through and clean-up Gumroad and MyMiniFactory websites once we merge and deploy all this.

Just pushed a fix for some MyMiniFactory issues I was seeing. I still saw some duplicated SourceSites for MyMiniFactory, but I'm not sure if that was due to me messing up the DB, the SourceSite itself, or the submission handler. I'm going to reset the PPE DB to a recent backup to check.

**Comment by aluhrs13** [comment]
Did a diff of the production and preproduction databases to figure out more about #119, it looks like MiniSourceSite.Link is NULL for all Gumroad Minis that were in the db pre-migration. Do we need to do something like what you did for MyMiniFactory but for Gumroad?

**Comment by aluhrs13** [comment]
Only remaining open issues I have in my head right now:
- [ ] MiniSourceSite.Link for Gumroad are NULL
- [ ] Need to re-enable Facebook (mostly checklist for finalizing the PR)
- [ ] Trying to add a Mini from a creator's page (/Minis/Create?URL=) doesn't work.
I'll try to do another thorough test pass in the next day or so.

**Comment by anaximander23** [comment]
The reason that adding a mini from a creator's page doesn't work is because it's making a GET request; technically speaking a GET shouldn't have side effects like creating things in the database. One simple way to make this work would be to make the button a form submit so that it can POST instead.

**Pull request merged**